client-core/src/service/hdkey_service.rs

```diff
@@ -0,0 +1,184 @@
+use bip39::{Language, Mnemonic, MnemonicType, Seed};
+use secstr::SecUtf8;
+use zeroize::Zeroize;
+
+use client_common::{PrivateKey, PublicKey, Result, SecureStorage, Storage};
+use log::debug;
+use tiny_hderive::bip32::ExtendedPrivKey;
+const KEYSPACE: &str = "core_hdkey";
+use super::key_service_data::KeyServiceInterface;
+
+/// Maintains mapping `public-key -> private-key`
+#[derive(Debug, Default, Clone)]
+pub struct HDKeyService<T: Storage> {
+    storage: T,
+}
+
+impl<T> KeyServiceInterface for HDKeyService<T>
+where
+    T: Storage,
+{
+    /// m / purpose' / coin_type' / account' / change / address_index
+    /// account: donation, savings, common expense
+    /// change: 0: external, 1: internal
+    /// Generates a new public-private keypair
+    fn generate_keypair(
+        &self,
+        name: &str,           // wallet name
+        passphrase: &SecUtf8, // wallet pass phrase
+        is_staking: bool,     // kind of address
+    ) -> Result<(PublicKey, PrivateKey)> {
+        let seed_bytes = self.storage.get_secure(KEYSPACE, name, passphrase)?;
+        let mut index = if is_staking {
+            self.read_number(passphrase, format!("staking_{}", name).as_bytes(), 0)
+        } else {
+            self.read_number(passphrase, format!("transfer_{}", name).as_bytes(), 0)
+        };
+        debug!("hdwallet index={}", index);
+        let account = if is_staking { 1 } else { 0 };
+        let extended = ExtendedPrivKey::derive(
+            &seed_bytes.clone().unwrap(),
+            format!("m/44'/394'/{}'/0/{}", account, index).as_str(),
```

Comment:
394 is main net only... for testnet / devnet, it should use 1 / 0x80000001

Reply:
ok